### PR TITLE
Implement "startup-time" type checking 🎉

### PR DESCRIPTION
**Pull request by hirrolot**
This PR introduces the `dptree::type_check` function that makes sure that all types required by `Handler` are present in `DependencyMap`. In order to decide which types are required by `Handler`, we use (run-time) type inference.

While the changes in this PR are formally **BC**, little or no intervention is required by a `dptree` user to update their code. The main addition is `dptree::type_check`, which is designed to be called _before_ executing a handler; for example, in `teloxide`, we could invoke `dptree::type_check` at some point while building `Dispatcher`. Given that, practically everywhere, a dispatch tree is constructed at a program startup, our type checking should be almost the same as static type checking (modulo different type mismatch messages).

See more information in `CHANGELOG.md`.

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
It's great to finally see this issue being addressed in the dptree, thanks for working on it!

I'll try to review soon™, although the changes look substantial, so it can take some time.

**Comment by Hirrolot** [comment]
Thanks for your time; when reviewing, it can be easier to first look at the unit tests from `handler/core.rs` to understand the type checking/inference logic:
 - https://github.com/teloxide/dptree/blob/d8663dd16f05ae83c90b661539bee529c427c389/src/handler/core.rs#L842
 - https://github.com/teloxide/dptree/blob/d8663dd16f05ae83c90b661539bee529c427c389/src/handler/core.rs#L902

File: README.md [hunk 1]
```diff
@@ -100,7 +100,9 @@ We decided to use a [continuation-passing style (CPS)] internally and expose nea
...
+In Rust, it is possible to express type-safe DI that checks all types statically. However, this would require complex type-level manipulations (like those in the [frunk] library). [@p0lunin] and I ([@Hirrolot]) decided not to trade comprehensible error messages for compile-time safety, since we had a plenty of experience that the uninitiated users simply cannot understand what is wrong with their code, owing to the utterly inadequate diagnostic messages from rustc.
+
+The approach taken by `dptree` is to implement run-time type checking instead, via `dptree::type_check`, to make sure that all required types are provided _before_ execution. If `dptree::type_check` is _not_ called, type checking will be delayed until execution (this is not recommended). This approach works much like static type checking when you build your whole dispatch tree at a program startup; the downside is that the panic message raised by `dptree::type_check` does not show the exact location to be fixed, only a final type mismatch message (the types in question are shown though!).
```
Comment (Hirrolot) [inline README.md:105]: By the way, I believe that the sentence saying that the panic message doesn't show "the exact location to be fixed" is not appropriate here. Even if the types were checked statically, we wouldn't be able to spot the location of an "error". Imagine that handler `F` is chained to handler `G`; `G` requires type `A` to be present, but `F` does not provide this type to `G`. Is this a type error or not? We can't decide because it might be the case that `F` should have provided this type or the whole resulting handler is meant to require type `A` to be present. This is why we don't type check inside `chain` and `branch`.

Comment (WaffleLapkin) [reply README.md:105]: We can still save the information of where the requirement comes from via [`Location`](https://doc.rust-lang.org/std/panic/struct.Location.html) this of course requires a more complicated structure, but IMO worth it.

File: src/handler/core.rs [hunk 2]
```diff
@@ -108,12 +162,47 @@ where
     pub fn chain(self, next: Self) -> Self {
         let required_update_kinds_set = self.description().merge_chain(next.description());
 
-        from_fn_with_description(required_update_kinds_set, move |event, cont| {
-            let this = self.clone();
-            let next = next.clone();
-
-            this.execute(event, |event| next.execute(event, cont))
-        })
+        let new_sig = match (&self.data.sig, &next.data.sig) {
```
Comment (WaffleLapkin) [inline src/handler/core.rs:165]: I'm a bit iffy about doing this here, could we maybe do all the type-checking via the descriptions (which should really have been named metadata, imo...)?

File: README.md (hunk 1 above)
Comment (Hirrolot) [reply README.md:105]: How are you suggesting to use `Location`? We don't even panic inside `.chain` or `.branch` (except the case where the second branch is `dptree::entry()`).

File: src/handler/core.rs (hunk 2 above)
Comment (Hirrolot) [reply src/handler/core.rs:165]: The reason we type-check not via descriptions is because we want to _always_ type-check handlers and still use descriptions. For example, in `teloxide` we're using `DpHandlerDescription`; to be able to type-check, we would need to somehow make the handler polymorphic over any description so as to instantiate the same handler with both `DpHandlerDescription` and "`TypeCheckDescription`".

Comment (WaffleLapkin) [reply src/handler/core.rs:165]: The descriptions are trivially composable though, you can include `TypeCheckDescription` inside `DpHandlerDescription` (just like we already use some description inside `DpHandlerDescription`, I don't remember the name tho...)

File: README.md (hunk 1 above)
Comment (WaffleLapkin) [reply README.md:105]: I propose that when we save "obligations" ("input to this handler must contain type `A`"), we also save where they are coming from (akin to rustc's [`ObligationCause`](https://github.com/rust-lang/rust/blob/91b61a4ad618c1abc2af43a58695de185ef1e513/compiler/rustc_middle/src/traits/mod.rs#L96))

File: src/handler/core.rs (hunk 2 above)
Comment (Hirrolot) [reply src/handler/core.rs:165]: But that would require a user to compose their description types with our `TypeCheckDescription` instead of type-checking automatically?

File: README.md (hunk 1 above)
Comment (Hirrolot) [reply README.md]: I've just pushed a commit addressing this issue. Thanks for showing a useful direction.

File: src/handler/core.rs [hunk 3]
```diff
@@ -194,52 +341,94 @@ where
     /// ```
     pub async fn execute<Cont, ContFut>(
         self,
-        container: Input,
+        input: Input,
```
[the remark this answers is not in the document]
Comment (Hirrolot) [reply src/handler/core.rs]: It has turned out to be indeed simpler. The `Input` generic was initially here when DI was optional, but since basically it's required now by almost all combinators, I don't see any reason to leave it generic. Probably one counterargument would be that the `Insert` and `DependencySupplier` traits make it explicit what operations are being invoked by particular functions, but I doubt that somebody really cares.

File: src/handler/core.rs [hunk 4]
```diff
@@ -108,12 +162,47 @@ where
...
-        from_fn_with_description(required_update_kinds_set, move |event, cont| {
-            let this = self.clone();
-            let next = next.clone();
-
-            this.execute(event, |event| next.execute(event, cont))
-        })
+        let new_sig = match (&self.data.sig, &next.data.sig) {
+            (_, HandlerRtSig::Entry) => {
+                panic!("Ill-typed handler chain: the second handler cannot be an entry")
+            }
```
[the remark this answers is not in the document]
Comment (Hirrolot) [reply src/handler/core.rs]: I don't know how I think about this. If the second handler is an entry, I'd say it's probably a programmer's mistake 99%.

**Comment by Hirrolot** [comment]
@WaffleLapkin, your suggestions have just been addressed.

File: src/handler/core.rs (hunk 2 above)
Comment (WaffleLapkin) [reply src/handler/core.rs:174]: Yes, although I'm not sure if it's a big deal?

We could also store the type checking description/meta always, e.g. `description: WithTypeCheck<Descr>`.

File: src/handler/core.rs [hunk 5]
```diff
@@ -508,4 +828,192 @@ mod tests {
...
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
+                output_types: hashset! {},
+                obligations: hashmap! { /* Must not be used. */ },
+            },
+            &deps,
+        );
+    }
+
+    #[test]
+    // Cannot specify the exact panic message because `HashSet` is iterated in a random order.
```
Comment (WaffleLapkin) [inline src/handler/core.rs:870]: We should probably use deterministic hash builder or index set.

File: src/handler/core.rs [hunk 6]
```diff
@@ -508,4 +828,192 @@ mod tests {
...
+        #[derive(Clone)]
+        struct C;
+
+        let input_types = vec![Type::of::<A>(), Type::of::<B>(), Type::of::<C>()];
+
+        type_check(
+            &HandlerSignature::Other {
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
+                output_types: hashset! {},
+                obligations: hashmap! { /* Must not be used. */ },
```
Comment (WaffleLapkin) [inline src/handler/core.rs:886]: What does `/* Must not be used */` means?

File: src/handler/core.rs [hunk 7]
```diff
@@ -508,4 +828,192 @@ mod tests {
...
+        #[derive(Clone)]
+        struct B;
+        #[derive(Clone)]
+        struct C;
+
+        let input_types = vec![Type::of::<A>(), Type::of::<B>(), Type::of::<C>()];
+
+        type_check(
+            &HandlerSignature::Other {
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
```
Comment (WaffleLapkin) [inline src/handler/core.rs:884]: This could just be `HashSet::from([Type::of::<A>(), Type::of::<B>(), Type::of::<C>()])`

Comment (Hirrolot) [reply src/handler/core.rs]: Fixed.

File: src/handler/endpoint.rs [hunk 8]
```diff
@@ -1,35 +1,49 @@
...
+        move |x, _cont| {
+            let f = Arc::clone(&f);
+            async move {
+                let f = f.inject(&x);
+                f().map(ControlFlow::Break).await
+            }
+        },
+        HandlerSignature::Other {
+            input_types: F::input_types(),
+            output_types: HashSet::from([]),
```
[the remark this answers is not in the document]
Comment (Hirrolot) [reply src/handler/endpoint.rs]: Fixed.

File: src/handler/core.rs (hunk 6 above)
Comment (Hirrolot) [reply src/handler/core.rs:884]: `obligations` are accessed by `type_check` only for printing purposes, when a typing error occurs. Since there is no error here, we can just leave `obligations` empty.

File: src/di.rs [hunk 9]
```diff
@@ -172,27 +146,49 @@ where
...
 
-impl<V, S> DependencySupplier<V> for Arc<S>
-where
-    S: DependencySupplier<V>,
-{
-    fn get(&self) -> Arc<V> {
-        self.deref().get()
+impl Debug for DependencyMap {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        f.debug_struct("DependencyMap").finish()
```
[the remark this answers is not in the document]
Comment (Hirrolot) [reply src/di.rs:163]: I don't remember why a manual implementation of `Debug` was chosen. Maybe just `#[derive(Debug)]`?

File: src/di.rs [hunk 10]
```diff
@@ -172,27 +146,49 @@ where
             .downcast::<V>()
             .expect("Checked by .unwrap_or_else()")
     }
+
+    fn available_types(&self) -> String {
```
[the remark this answers is not in the document]
Comment (Hirrolot) [reply src/di.rs:150]: There is only one caller though. Returning an iterator of type names would be just `self.map.values().map(|dep| dep.type_name)` and would put the burden of formatting upon a caller.

File: src/handler/core.rs (hunk 2 above)
Comment (Hirrolot) [reply src/handler/core.rs:174]: I just think that type checking is more "inherent" to `Handler` than a typical description. For example, we explicitly construct `HandlerSignature` independently of descriptions in `endpoint`, `filter`, `inspect`, etc., because that requires access to `Injectable`'s generics that `HandlerDescription` simply doesn't have access to. Maybe we could somehow circumvent such limitations but I don't think it'd result in a simpler structure...

File: src/handler/core.rs (hunk 5 above)
Comment (Hirrolot) [reply src/handler/core.rs:870]: Do you know if it's possible with the standard library's `HashSet` or we should use another collection?

**Comment by Hirrolot** [comment]
Ping @WaffleLapkin

**Comment by WaffleLapkin** [comment]
I'm still having very limited time/energy rn, having a hard time IRL. Sorry for not reviewing this for so long...

**Comment by Hirrolot** [comment]
@WaffleLapkin, do you have plans reviewing this PR? It seems that it's most requested by our users.

File: src/handler/core.rs [hunk 11]
```diff
@@ -44,36 +52,71 @@ use crate::{description, HandlerDescription};
...
+    Other {
+        /// The set of types that the handler accepts. In the case of a DI
+        /// container, this is the set of types that this handler retrieves from
+        /// the container.
+        input_types: HashSet<Type>,
+        /// The set of types that this handler passes down the chain.
+        output_types: HashSet<Type>,
+        /// The set of "obligations", i.e., code locations where specific input
+        /// types are required.
+        obligations: HashMap<Type, &'static Location<'static>>,
```
Comment (WaffleLapkin) [inline src/handler/core.rs:83]: You don't need `input_types` here, because the same role is already fulfilled by `obligations`.

File: src/handler/core.rs (hunk 2 above)
Comment (WaffleLapkin) [reply src/handler/core.rs:174]: Yeah okay, it's fine.

File: src/handler/core.rs [hunk 12]
```diff
@@ -108,12 +171,63 @@ where
...
 
-        from_fn_with_description(required_update_kinds_set, move |event, cont| {
-            let this = self.clone();
-            let next = next.clone();
-
-            this.execute(event, |event| next.execute(event, cont))
-        })
+        let new_sig = match (&self.data.sig, &next.data.sig) {
+            (_, HandlerSignature::Entry) => {
+                panic!("Ill-typed handler chain: the second handler cannot be an entry")
```
Comment (WaffleLapkin) [inline src/handler/core.rs:176]: Why? `entry` as rhs should be trivially fine.

File: src/handler/core.rs [hunk 13]
```diff
@@ -108,12 +171,63 @@ where
...
+            }
+            (HandlerSignature::Entry, HandlerSignature::Other { .. }) => next.data.sig.clone(),
```
Comment (WaffleLapkin) [inline src/handler/core.rs:178]:
```suggestion
            (HandlerSignature::Entry, other @ HandlerSignature::Other { .. }) => other.clone(),
```

File: src/handler/core.rs [hunk 14]
```diff
@@ -122,6 +236,25 @@ where
     /// chaining and
     /// branching"](#the-difference-between-chaining-and-branching).
     ///
+    /// # Type inference algorithm
```
Comment (WaffleLapkin) [inline src/handler/core.rs:239]: You should replace all mentions of "inference" with "checking", we don't actually do any inference.

File: src/handler/core.rs [hunk 15]
```diff
@@ -159,17 +292,61 @@ where
...
+                    // Since any of the two handlers can end up being executed, take the union of
+                    // the input types of both handlers.
+                    input_types: self_input_types.union(next_input_types).cloned().collect(),
+
+                    // Since any of the two handlers can end up being executed, take the
+                    // intersection of the output types of both handlers.
+                    output_types: next_output_types
+                        .intersection(self_output_types)
+                        .cloned()
+                        .collect(),
```
Comment (WaffleLapkin) [inline src/handler/core.rs:322]: that doesn't sound right? I `.branch` is only supposed to output things from self, not next

File: src/handler/core.rs [hunk 16]
```diff
@@ -44,36 +52,71 @@ use crate::{description, HandlerDescription};
...
+        /// types are required.
+        obligations: HashMap<Type, &'static Location<'static>>,
+    },
+}
+
+/// A run-time representation of a type. Used only for run-time type inference
+/// and checking of handler chains.
+///
+/// See [`crate::type_check`].
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
```
Comment (WaffleLapkin) [inline src/handler/core.rs:91]: `Hash` impl should only really consider `id`, there is no reason to hash `name`, since it is always the same for a given `id`. `PartialEq` impl should also probably have a fast path for `ptr::eq(a.name, b.name)`.

Comment (WaffleLapkin) [reply src/handler/core.rs:91]: (ideally we'd only have `TypeId` as a key, and had name in the _value_, but if it's simpler to have `Type` as a key, let's make it a bit nicer)

File: src/handler/core.rs [hunk 17]
```diff
@@ -44,36 +52,71 @@ use crate::{description, HandlerDescription};
...
-type DynF<'a, Input, Output> =
-    dyn Fn(Input, Cont<'a, Input, Output>) -> HandlerResult<'a, Input, Output> + Send + Sync + 'a;
+/// A run-time handler signature. Used only for run-time type inference and
+/// checking of handler chains.
+///
+/// See [`crate::type_check`].
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HandlerSignature {
+    /// The signature corresponding to [`crate::entry`].
+    Entry,
```
Comment (WaffleLapkin) [inline src/handler/core.rs:72]: why is this needed at all? isn't `entry()`, for purposes of type checking DI just a `input_types = [], obligations = {}`?

File: src/handler/core.rs [hunk 18]
```diff
@@ -258,32 +483,93 @@ where
...
+    match sig {
+        HandlerSignature::Entry => {}
+        HandlerSignature::Other { input_types, output_types: _, obligations } => {
+            let container_types = container
+                .map
+                .iter()
+                .map(|(type_id, dep)| Type { id: *type_id, name: dep.type_name })
+                .collect::<HashSet<_>>();
+
+            if !input_types.is_subset(&container_types) {
```
Comment (WaffleLapkin) [inline src/handler/core.rs:544]: I think doing `input_types.difference(&container_types)` from the start and then  checking if it's empty is a bit nicer

File: src/handler/core.rs (hunk 5 above)
Comment (WaffleLapkin) [reply src/handler/core.rs:870]: You can use `rustc-hash` crate (it provides a hasher that can be used with the standard library collections)

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]

File: src/di.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply src/di.rs:150]: That was what I wanted (caller may have different formatting needs). But since this is only called once and is a private helper, ig that's fine.

File: src/di.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply src/di.rs:163]: You can't just `derive(Debug)`, because `dyn Any + Send + Sync` does not implement `Debug`.

Although mh. You should probably derive it on `DependencyMap` and then do a custom impl for `Dependency`.

File: src/handler/core.rs (hunk 5 above)
Comment (hirrolot) [reply src/handler/core.rs]: Done.

File: src/handler/core.rs (hunk 14 above)
Comment (hirrolot) [reply src/handler/core.rs:237]: We do dynamic type inference inside `chain` and `branch` and checking inside `type_check`, but it's not unification-based type inference, of course.

File: src/handler/core.rs (hunk 13 above)
Comment (hirrolot) [reply src/handler/core.rs]: Done.

File: src/di.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (hirrolot) [reply src/di.rs]: Done.

File: src/handler/core.rs (hunk 16 above)
Comment (hirrolot) [reply src/handler/core.rs]: I don't understand why would we need `ptr::eq` on type names instead of just comparing type IDs, which are just two machine words in size. Otherwise, done.

File: src/handler/core.rs (hunk 18 above)
Comment (hirrolot) [reply src/handler/core.rs:568]: How "from the start"? In this function, we compute `container_types` and then immediately check if `input_types` is a subset of `container_types`.

File: src/handler/core.rs (hunk 15 above)
Comment (hirrolot) [reply src/handler/core.rs:346]: Doesn't the comment clarify this situation? Any of the two handlers can be executed, so we take the intersection of their output types as an approximation.

File: src/handler/core.rs (hunk 17 above)
Comment (hirrolot) [reply src/handler/core.rs:78]: Then we would need `output_types` to be the set of all types, because otherwise, an intersection of any sequence of branches with an entry branch would be always empty.

File: src/handler/core.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hirrolot) [reply src/handler/core.rs]: I still think so. We can remove this restriction in the future if needed.

File: src/handler/core.rs (hunk 12 above)
Comment (hirrolot) [reply src/handler/core.rs:196]: Answered in https://github.com/teloxide/dptree/pull/23#discussion_r1199691364.

File: src/handler/core.rs (hunk 11 above)
Comment (hirrolot) [reply src/handler/core.rs]: Done.
